unison-core/src/Unison/ConstructorReference.hs

```diff
@@ -0,0 +1,35 @@
+-- | The constructor reference type.
+module Unison.ConstructorReference
+  ( GConstructorReference (..),
+    ConstructorReference,
+    ConstructorReferenceId,
+    reference_,
+    toShortHash,
+  )
+where
+
+import Control.Lens
+import Unison.DataDeclaration.ConstructorId (ConstructorId)
+import Unison.Prelude
+import Unison.Reference (TypeReference, TypeReferenceId)
+import qualified Unison.Reference as Reference
+import Unison.ShortHash (ShortHash)
+import qualified Unison.ShortHash as ShortHash
+
+-- | A reference to a constructor is represented by a reference to its type declaration, plus the ordinal constructor id.
+data GConstructorReference r
+  = ConstructorReference !r !ConstructorId
+  deriving stock (Eq, Functor, Ord, Show)
+
+type ConstructorReference = GConstructorReference TypeReference
+
+type ConstructorReferenceId = GConstructorReference TypeReferenceId
```

Comment:
The other modules just call this `Id` and expect qualified usage, which I think makes sense. What do you think of doing that here?

Reply:
I think that's a reasonable idea for consistency, but what to do about these?

```haskell
-- in Unison.Reference
type TypeReferenceId = Id
type TermReferenceId = Id
```

It'd be a bigger change to go the other direction and not name things e.g. `Reference.Id`, `Referent.Id`, but that's my preference, because (for one thing) haddocks become a lot more readable.

Reply:
Good question, I wonder if eventually it might make sense to merge everything into the same module if we can get good names for each concept, then you can just import the whole thing qualified and use `R.TypeReferenceId` etc.

For now whatever you decide is fine with me 👍🏼